### ISSUE TITLE
feature req: pre-built binaries with CGO and libjpeg-turbo

### ISSUE THREAD

**Issue by msageryd** [klippa-app/pdfium-cli#15]
I'm creating a separate issue for this so it's easier to track. The original thread (#2) mentioning this is way to long.

The build command for building a blazingly fast version of pdfium-cli is simple:
`go build -tags pdfium_cli_use_cgo,pdfium_use_turbojpeg`

What turned out being not so simple was to have the compiler/linker to find the header files and lib files for `pdfium` and `libjpeg-turbo`. I simply didn't succeed with turbojpeg.

It would be awesome to be able to download pre-built binaries for Mac and Linux. Others might want Windows binaries as well.

**Comment by jerbob92** [comment]
Thanks! I was already working on this here: #13 

You want to deploy on Alpine containers right? By default these don't have libc right? This means I probably have to built with musl, however, it seems that there currently is an issue with the musl pdfium build: https://github.com/bblanchon/pdfium-binaries/issues/129

I will let you know once I figured that one out.

**Comment by msageryd** [comment]
I will split my Docker build as you suggested earlier, i.e. build stuff in one image and then copy the essentials to a clean Alpine image. Hence I don't mind installing the build tools needed, as long as they are available via `apk`.

Currently I have `alpine-sdk` and `pkgconfig` installed. I don't know if alpine-sdk includes libc.

**Comment by jerbob92** [comment]
I think libc is something that you would also need during runtime, so I don't think splitting up the build steps would fix that. Another issue with using libc would be that it's quite version specific, so a build made from Ubuntu 22.04 might not work on 20.04, and probably also not on Alpine, musl doesn't have this issue afaik, so I think musl should be the way to go.

**Comment by jerbob92** [comment]
Yesterday I tried to resolve the musl build issues in pdfium, but no luck so far, will keep you posted.

**Comment by msageryd** [comment]
Alpine is nice to work with since the containers deploy and spin up fast. But If there is too much problem with Alpine I could use Ubuntu instead.

**Comment by jerbob92** [comment]
Yeah the musl build should definitly work, I also want to build pdfium-cli with musl so that the (Webassembly) linux build works on more linux systems. It's also in the build list of https://github.com/bblanchon/pdfium-binaries so it should be fixed.

**Comment by jerbob92** [comment]
I have probably found a fix and made a PR on pdfium-binaries. I compiled pdfium-cli + pdfium + libjpegturbo with musl now and put it all together in a pure docker Alpine container, so the Dockerfile looked something like this:

```
FROM alpine:latest
COPY main /main
COPY libturbojpeg.so /libturbojpeg.so.0
COPY libpdfium.so /libpdfium.so
ENV LD_LIBRARY_PATH="/"
ENTRYPOINT ["/main"]
```

Based on these results, I'm pretty sure this will work once the PR has been merged into pdfium-binaries.

**Comment by msageryd** [comment]
I'm not sure how far you've got with this. Should I find new binaries along with the 0.4 release, or are you still waiting for the pdfium merge?

**Comment by jerbob92** [comment]
On my side everything is ready to go, I have added tests and build steps for the CGO/libjpeg-turbo binaries: https://github.com/klippa-app/pdfium-cli/actions/runs/6344835852

However, the musl build is still failing. My PR has been merged in (https://github.com/bblanchon/pdfium-binaries/pull/130), but it's waiting for a new release now. pdfium-binaries automatically build each week, so it will take a few more days for the new version to be available.

**Comment by msageryd** [comment]
Sounds great. 
No hurry, I just wanted to make sure I hadn't missed anything.

Again, thank you soo much for your efforts. The file processing chain I'm building now would not have been nearly as effective with other tools. Your multi-file-stream on stdout makes all the difference.

**Comment by jerbob92** [comment]
The alpine build has been replaced for us with a proper one! It looks like the musl build of turbojpeg still has a symbol error, so I will have to look into that, should be solved soon.

**Comment by jerbob92** [comment]
This has been released in version [v0.4.1](https://github.com/klippa-app/pdfium-cli/releases/tag/v0.4.1).

This is the dockerfile I tested it with: 
```
FROM alpine:edge

RUN apk add libjpeg-turbo-dev curl

RUN curl -L https://github.com/bblanchon/pdfium-binaries/releases/download/chromium%2F6029/pdfium-linux-musl-x64.tgz -o pdfium-linux-musl-x64.tgz && \
    mkdir /opt/pdfium && \
    tar -C /opt/pdfium -xvf pdfium-linux-musl-x64.tgz && \
    rm pdfium-linux-musl-x64.tgz

RUN curl -L https://github.com/klippa-app/pdfium-cli/releases/download/v0.4.1/pdfium-native-linux-musl-amd64 -o /opt/pdfium-cli && chmod a+x /opt/pdfium-cli

ENV LD_LIBRARY_PATH=/opt/pdfium/lib:/usr/lib

CMD ["/opt/pdfium-cli"]
```
